go/vt/vtgate/planbuilder/operator_transformers.go

```diff
@@ -59,17 +61,70 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
 		return transformCorrelatedSubQueryPlan(ctx, op)
 	case *operators.Derived:
 		return transformDerivedPlan(ctx, op)
-	case *operators.SimpleProjection:
-		return transformSimpleProjection(ctx, op)
 	case *operators.Filter:
 		return transformFilter(ctx, op)
 	case *operators.Horizon:
 		return transformHorizon(ctx, op, isRoot)
+	case *operators.Projection:
+		return transformProjection(ctx, op)
 	}
 
 	return nil, vterrors.VT13001(fmt.Sprintf("unknown type encountered: %T (transformToLogicalPlan)", op))
 }
 
+func transformProjection(ctx *plancontext.PlanningContext, op *operators.Projection) (logicalPlan, error) {
+	src, err := transformToLogicalPlan(ctx, op.Source, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if cols := op.AllOffsets(); cols != nil {
+		// if all this op is doing is passing through columns from the input, we
+		// can use the faster SimpleProjection
+		return simplifyPlan(op, cols, src)
+	}
+
+	expressions := slices2.Map(op.Columns, func(from operators.ProjExpr) sqlparser.Expr {
+		return from.GetExpr()
+	})
+
+	return &projection{
+		source:      src,
+		columnNames: op.ColumnNames,
+		columns:     expressions,
+	}, nil
+}
+
+func simplifyPlan(op *operators.Projection, cols []int, src logicalPlan) (logicalPlan, error) {
+	// all columns are just passing through something from below
+	columns, err := op.Source.GetColumns()
+	if err != nil {
+		return nil, err
+	}
+	if len(columns) == len(cols) && elementsMatchIndices(cols) {
+		// the columns are already in the right order. we don't need anything at all here
+		return src, nil
+	}
+	return &simpleProjection{
+		logicalPlanCommon: newBuilderCommon(src),
+		eSimpleProj: &engine.SimpleProjection{
+			Cols: cols,
+		},
+	}, nil
+}
+
+// elementsMatchIndices checks if the elements of the input slice match
+// their corresponding index values. It returns true if all elements match
+// their indices, and false otherwise.
+func elementsMatchIndices(in []int) bool {
+	for idx, val := range in {
+		if val != idx {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
I think it would be nice to have some sort of `utils` package where we can put utility function like this one. Similar to the `utils` end-to-end package.

Reply:
This would force us to reuse code we already have and, not saying this is duplicated, but it would avoid duplication in the future and ease maintenance / testability